nano/node/socket.cpp

```diff
@@ -0,0 +1,334 @@
+
+	boost::asio::dispatch (strand, [this_l]() {
+		this_l->close_internal ();
+		this_l->acceptor.close ();
+		for (auto & connection_w : this_l->connections)
+		{
+			if (auto connection_l = connection_w.lock ())
+			{
+				connection_l->close ();
+			}
+		}
+		this_l->connections.clear ();
+	});
+}
+
+void nano::server_socket::on_connection (std::function<bool(std::shared_ptr<nano::socket>, boost::system::error_code const &)> callback_a)
+{
+	auto this_l (std::static_pointer_cast<nano::server_socket> (shared_from_this ()));
+
+	boost::asio::post (strand, [this_l, callback_a]() {
+		if (auto node_l = this_l->node.lock ())
+		{
+			if (this_l->acceptor.is_open ())
+			{
+				if (this_l->connections.size () < this_l->max_inbound_connections)
+				{
+					// Prepare new connection
+					auto new_connection (std::make_shared<nano::socket> (node_l->shared (), boost::none, this_l->concurrency_new_connections));
+
+					this_l->acceptor.async_accept (new_connection->tcp_socket, new_connection->remote,
+					boost::asio::bind_executor (this_l->strand,
+					[this_l, new_connection, callback_a](boost::system::error_code const & ec_a) {
+						if (auto node_l = this_l->node.lock ())
+						{
+							if (!ec_a)
+							{
+								// Make sure the new connection doesn't idle. Note that in most cases, the callback is going to start
+								// an IO operation immediately, which will start a timer.
+								new_connection->start_timer (node_l->config.tcp_idle_timeout);
+								node_l->stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_success, nano::stat::dir::in);
+								this_l->connections.push_back (new_connection);
+								this_l->evict_dead_connections ();
+							}
+							else
+							{
+								node_l->logger.try_log ("Unable to accept connection: ", ec_a.message ());
+							}
+
+							// If the callback returns true, keep accepting new connections
+							if (callback_a (new_connection, ec_a))
+							{
+								this_l->on_connection (callback_a);
+							}
+							else
+							{
+								node_l->logger.try_log ("Stopping to accept connections");
+							}
+						}
+					}));
+				}
+				else
+				{
+					this_l->evict_dead_connections ();
+					node_l->stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+					this_l->deferred_accept_timer.expires_after (std::chrono::seconds (2));
+					this_l->deferred_accept_timer.async_wait ([this_l, callback_a](const boost::system::error_code & ec_a) {
+						if (!ec_a)
+						{
+							// Try accepting again
+							std::static_pointer_cast<nano::server_socket> (this_l)->on_connection (callback_a);
+						}
+						else
+						{
+							if (auto node_l = this_l->node.lock ())
+							{
+								node_l->logger.try_log ("Unable to accept connection (deferred): ", ec_a.message ());
+							}
+						}
+					});
+				}
+			}
+		}
+	});
+}
+
+// This must be called from a strand
```

Comment:
Can this be checked programatically in an assert?

Reply:
Same as above.